spirv_glsl.cpp

```diff
@@ -349,12 +349,24 @@ void CompilerGLSL::emit_header()
 	auto &execution = get_entry_point();
 	statement("#version ", options.version, options.es && options.version > 100 ? " es" : "");
 
-	// Needed for binding = # on UBOs, etc.
-	if (!options.es && options.version < 420 && options.enable_420pack_extension)
+	bool can_use_image_load_store = true;
+
+	if (!options.es && options.version < 420)
 	{
-		statement("#ifdef GL_ARB_shading_language_420pack");
-		statement("#extension GL_ARB_shading_language_420pack : require");
-		statement("#endif");
+		// Needed for binding = # on UBOs, etc.
```

Comment:
I think it makes more sense to only enable the shader_image_load_store extension if the extension is actually required. You know this because you only need it if ExecutionModeEarlyFragmentTests is actually enabled, and this can be checked here. Also, there can be double declaration, because image_load_store extension will be placed in forced_extensions if image load store itself is used.

Propose change to check if early_fragment_tests is enabled and version < 420, then add image_load_store to forced_extensions.

Reply:
Thank you for pointing me towards `forced_extensions`, I hadn't spotted that.

So I'm trying to do two things here:
1) Make sure that the code emitted by SPIRV-Cross actually generates compilable code for versions < 420 and the use of `EarlyFragmentTests`. I believe this is now correctly done with this updated PR.
2) Make shaders that use `early_fragment_tests` usable on machines that do not support the extension, nor support OpenGL 420, by stripping the `early_fragment_tests` qualifier and any explicit requirement of the extension. I'm guessing this should be done as a transform on the SPIRV, and not as an option in SPIRV-Cross?

Reply:
You can use the unset_execution_mode() or whatever it is called to unset the early_fragment_tests execution mode. This way, the shader can be compat with older shader versions without introducing more options.

Reply:
Perfect - as we're using this as a library, I can call this function directly and I do not require any other changes. Thank you.

Reply:
(That said, this PR still fixes a bug, which I'd like to fix)

Reply:
I assume the bug you are referring to is that use of EarlyFragmentTests should enable image_load_store extension on < 420 targets? If so, yes, I'm happy to merge that.

Reply:
> I assume the bug you are referring to is that use of EarlyFragmentTests should enable image_load_store extension on < 420 targets?

Yes. Please let me know if there are any more changes you'd like from this PR.

Reply:
This PR should only do that then, no option for image load store extension should be there.

Reply:
I force-pushed the branch, as the roll-back was significant.
Can I confirm you're seeing the latest PR?

Reply:
I can see it now.

Reply:
Will merge once it passes Travis.

Reply:
Great, thank you.